cpp/src/arrow/gpu/cuda_context.cc

```diff
@@ -60,26 +69,32 @@ class CudaContext::CudaContextImpl {
   int64_t bytes_allocated() const { return bytes_allocated_.load(); }
 
   Status Allocate(int64_t nbytes, uint8_t** out) {
+    CU_RETURN_NOT_OK(cuCtxGetCurrent(&prev_context_));
     CU_RETURN_NOT_OK(cuCtxSetCurrent(context_));
 
     CUdeviceptr data;
     CU_RETURN_NOT_OK(cuMemAlloc(&data, static_cast<size_t>(nbytes)));
     bytes_allocated_ += nbytes;
     *out = reinterpret_cast<uint8_t*>(data);
+    if (prev_context_ != NULL) CU_RETURN_NOT_OK(cuCtxSetCurrent(prev_context_));
```

Comment:
This almost feels like an RAII utility class is needed

Reply:
Indeed, otherwise the context won't be restored if `cuMemAlloc` fails above.

Reply:
Fixed by introducing `ContextSaver` class.